File: 2021W11/fienestar/pgr_17680.js
/*
    캐시
    https://programmers.co.kr/learn/courses/30/lessons/17680
*/
function solution(cacheSize, cities) {
    let cache = []
    let answer = 0
    
    if(cacheSize == 0)return 5 * cities.length
    
    for(let city of cities){
        city = city.toUpperCase()
        
        if(cache.includes(city)){
            cache.splice(cache.indexOf(city),1)
            answer += 1
        }else
            answer += 5
        
        if(cache.length == cacheSize)
            cache.shift()
        
        cache.push(city)
    }
    
    return answer;
}

File: 2021W11/violetxf/pgr_17680.cpp
/*
    캐시
    https://programmers.co.kr/learn/courses/30/lessons/17680
*/
#include <string>
#include <vector>
#include <algorithm>
#include <iostream>
using namespace std;
class Cache{
private:
    int size;
    int cursor;
    vector<string> caches;
public:
    Cache(int size){
        this->size = size;
        caches.resize(size);
    }
    bool push_new(string data){
        auto it = search(data);
        bool has = it!=caches.end();
        if(has){
            caches.erase(it);
        }else{
            caches.erase(caches.begin());
        }
        caches.push_back(data);
        return has;
    }
private:
    vector<string>::iterator search(string key){
        return find(caches.begin(), caches.end(), key);
        
    }
};
int solution(int cacheSize, vector<string> cities) {
    int answer = 0;
    if(cacheSize==0)return cities.size()*5;
    Cache cache(cacheSize);
    for(int i=0;i<cities.size();i++){
        string data = cities[i];
        transform(data.begin(), data.end(), data.begin(), [](unsigned char c){return tolower(c);});
        if(cache.push_new(data)){
            ++answer;
        } else {
            answer+=5;
        }
    }
    return answer;
}
